[UX] Add cluster event table to explain state changes to user

lloyd-brown opened the pull request:
<!-- Describe the changes in this PR -->

Added a cluster event table which keeps track of state transitions and events that happen. The goal of this is to provide more description to the user about why the cluster reaches certain states. This PR adds a column in sky status -v detailing the latest event that happened to a cluster to aid in debugging.

<!-- Describe the tests ran -->
<!-- Unit tests (tests/test_*.py) are part of GitHub CI; below are tests that launch on the cloud. -->

Tested (run the relevant ones):

- [x] Code formatting: install pre-commit (auto-check on commit) or `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `/smoke-test` (CI) or `pytest tests/test_smoke.py` (local)
- [ ] Relevant individual tests: `/smoke-test -k test_name` (CI) or `pytest tests/test_smoke.py::test_name` (local)
- [ ] Backward compatibility: `/quicktest-core` (CI) or `pytest tests/smoke_tests/test_backward_compat.py` (local)

This PR was tested by starting an AWS cluster and then terminating one of the worker nodes, verifying that the "LAST_EVENT" column in sky status -v showed the various stages of the cluster (provisioning, launching instances, successfully reaching UP, returning to INIT due to worker termination). The attached image shows the output at the end of this process where the cluster returned to INIT.

<img width="1720" height="88" alt="Screenshot 2025-08-08 at 11 26 00 AM" src="https://github.com/user-attachments/assets/57c18c20-674e-42a1-ae1c-d589eb31520d" />

The new column does cause line wrap in sky status -v which may not be ideal.

<!-- CI commands (/-prefixed) can only be triggered by repo members -->

SeungjinYang reviewed (commented):
This is great! I'm actually working on a feature that requires storing some kubernetes events in a database, and I'm hoping to use this new table created in the PR to do so. This is because apparently kubernetes events are only kept for one hour (see https://medium.com/@andrew.kaczynski/kubernetes-events-how-to-keep-historical-data-of-your-cluster-835d685cc45), so I'd need to persist them somewhere if I want a longer term retention of the events.

Now, I also see that you print the latest event that happened to a cluster in `sky status -v`, which I assume is only intended for state transitions. I'd ideally not want to overwrite the latest event for the cluster with some internal k8s event and have that show up, if possible.

How about we add a column in this new table that denotes the "type" of each event (for example, every event added in this PR could be of type "status_transition", and what I later add for k8s could be "debug" events, etc), so that the table could be used for multiple purposes without each stepping on another's toes?

lloyd-brown commented:
> This is great! I'm actually working on a feature that requires storing some kubernetes events in a database, and I'm hoping to use this new table created in the PR to do so. This is because apparently kubernetes events are only kept for one hour (see https://medium.com/@andrew.kaczynski/kubernetes-events-how-to-keep-historical-data-of-your-cluster-835d685cc45), so I'd need to persist them somewhere if I want a longer term retention of the events.
> 
> Now, I also see that you print the latest event that happened to a cluster in `sky status -v`, which I assume is only intended for state transitions. I'd ideally not want to overwrite the latest event for the cluster with some internal k8s event and have that show up, if possible.
> 
> How about we add a column in this new table that denotes the "type" of each event (for example, every event added in this PR could be of type "status_transition", and what I later add for k8s could be "debug" events, etc), so that the table could be used for multiple purposes without each stepping on another's toes?

Makes sense, I've added a "type" column and added an enum with two types STATUS_CHANGE and DEBUG. The last event strictly looks for the latest STATUS_CHANGE item now.

SeungjinYang commented on sky/global_user_state.py:
```diff
@@ -162,6 +163,33 @@
+cluster_event_table = sqlalchemy.Table(
+    'cluster_events',
+    Base.metadata,
+    sqlalchemy.Column('cluster_hash', sqlalchemy.Text, primary_key=True),
+    sqlalchemy.Column('name', sqlalchemy.Text),
+    sqlalchemy.Column('starting_status', sqlalchemy.Text),
+    sqlalchemy.Column('ending_status', sqlalchemy.Text),
+    sqlalchemy.Column('reason', sqlalchemy.Text, primary_key=True),
+    sqlalchemy.Column('transitioned_at', sqlalchemy.Integer, primary_key=True),
+    sqlalchemy.Column('type', sqlalchemy.Text),
```
On the primary key being (cluster_hash, reason) - would it deal with the case where, for example, a cluster was stopped and started repeatedly? The worry would be that, for example, two events with reason "Cluster was stopped by user.", for two different instances of cluster stoppage (after being restarted), would cause UNIQUE constraint failure.

SeungjinYang replied:
I see. I'm slightly worried this will lead to silent dropping of certain events. For example, if a cluster was launched -> stopped1 -> started1 -> stopped2 then it is possible that the last event (stopped2) will fail to commit to the database because a stopped event for the cluster already exists. This may mislead users into thinking the last event in the cluster (as displayed) is "started", when in reality it should've been "stopped".

I guess we can get around this my adding an autoincrementing ID column or something that can just distinguish between different inserted values. My question is, is there any reason to not prefer that approach - i.e. are there actually duplicate inserts that we'd like to catch, e.g. some record that gets added every time `sky status --refresh` is run?

SeungjinYang replied:
Actually now that we have the pkey on timestamp this should be fine. Thanks!

SeungjinYang commented:
One other thing we need to do when adding a new table (or changing the db schema in general) is to add a corresponding db migration, see https://github.com/skypilot-org/skypilot/pull/6554/files (specifically sky/schemas/db/global_user_state) for an example

lloyd-brown commented:
> One other thing we need to do when adding a new table (or changing the db schema in general) is to add a corresponding db migration, see https://github.com/skypilot-org/skypilot/pull/6554/files (specifically sky/schemas/db/global_user_state) for an example

Added a migration!

SeungjinYang commented on sky/schemas/db/global_user_state/005_cluster_event.py:
```diff
@@ -0,0 +1,61 @@
+        # Add new table for cluster events.
+        db_utils.add_tables_to_db_sqlalchemy(Base.metadata, op.get_bind())
+
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'cluster_hash',
+                                             sa.Text(),
+                                             server_default=None)
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'name',
+                                             sa.Text(),
+                                             server_default=None)
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'starting_status',
+                                             sa.Text(),
+                                             server_default=None)
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'ending_status',
+                                             sa.Text(),
+                                             server_default=None)
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'reason',
+                                             sa.Text(),
+                                             server_default=None)
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'transitioned_at',
+                                             sa.Integer(),
+                                             server_default=None)
+        db_utils.add_column_to_table_alembic('cluster_events',
+                                             'type',
+                                             sa.Text(),
+                                             server_default=None)
```
Since we're adding the table for the first time, we don't need these statements (these statements are for when new columns are added to an existing table)

SeungjinYang commented on sky/schemas/db/global_user_state/005_cluster_event.py:
```diff
@@ -0,0 +1,61 @@
+    """Add new table for cluster events."""
+    with op.get_context().autocommit_block():
+        # Add new table for cluster events.
+        db_utils.add_tables_to_db_sqlalchemy(Base.metadata, op.get_bind())
```
What we'd like to do here ideally is to create just the new table introduced in the PR, as the migration is supposed to be the diff between the previous db state and the next db state. I am adding a function in https://github.com/skypilot-org/skypilot/pull/6594 to allow a specific table to be created, which you can use after its merge here.

lloyd-brown commented:
/smoke-test --aws test_launch_fast_with_autostop

lloyd-brown commented:
/smoke-test -k test_launch_fast_with_autostop --aws

lloyd-brown commented:
/smoke-test -k test_managed_jobs_recovery_aws --aws
/smoke-test -k test_managed_jobs_pipeline_recovery_aws --aws
/smoke-test -k test_managed_jobs_cancellation_aws --aws

kevinmingtarja commented:
/smoke-test -k test_autodown --kubernetes
